subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1303 @@
+				return;
+		}
+
+		connection_state = STATE_CONNECTED;
+
+		LOG_DBG("MQTT client connected");
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+		if (dps_reg_in_progress()) {
+			err = dps_subscribe();
+			evt.type = AZURE_IOT_HUB_EVT_DPS_REGISTERING;
+		} else
+#endif
+		{
+			err = topic_subscribe();
+			evt.type = AZURE_IOT_HUB_EVT_CONNECTED;
+		}
+
+		if (err) {
+			LOG_ERR("Failed to subscribe to topics, err: %d", err);
+		}
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_DISCONNECT:
+		LOG_DBG("MQTT_EVT_DISCONNECT: result = %d", mqtt_evt->result);
+
+		connection_state = STATE_INIT;
+		evt.type = AZURE_IOT_HUB_EVT_DISCONNECTED;
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+		if (dps_reg_in_progress()) {
+			if (dps_process_message(&evt)) {
+				/* The message has been processed */
+				break;
+			}
+		}
+#endif
+		if (atomic_get(&dps_disconnecting) == 1) {
+			break;
+		}
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_PUBLISH:
+		on_publish(client, mqtt_evt);
+		break;
+	case MQTT_EVT_PUBACK:
+		LOG_DBG("MQTT_EVT_PUBACK: id = %d result = %d",
+			mqtt_evt->param.puback.message_id,
+			mqtt_evt->result);
+
+		if (dps_reg_in_progress()) {
+			break;
+		}
+
+		evt.type = AZURE_IOT_HUB_EVT_PUBACK;
+		evt.data.message_id = mqtt_evt->param.puback.message_id;
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_SUBACK:
+		LOG_DBG("MQTT_EVT_SUBACK: id = %d result = %d",
+			mqtt_evt->param.suback.message_id,
+			mqtt_evt->result);
+
+		if (on_suback()) {
+			/* Event DPS-related, and should not be propagated */
+			break;
+		}
+
+		evt.type = AZURE_IOT_HUB_EVT_READY;
+		azure_iot_hub_notify_event(&evt);
+		break;
+	default:
+		break;
+	}
+}
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+static struct mqtt_utf8 *user_name_get(void)
+{
+	ssize_t len;
+
+	len = snprintk(user_name_buf, sizeof(user_name_buf), USER_NAME_TEMPLATE,
+		       dps_hostname_get(), conn_config.device_id);
+	if ((len < 0) || (len > sizeof(user_name_buf))) {
+		LOG_ERR("Failed to create user name");
+		return NULL;
+	}
+	user_name.size = len;
+	return &user_name;
+}
+#endif
+
+#if defined(CONFIG_AZURE_IOT_HUB_STATIC_IPV4)
+static int broker_init(bool dps)
+{
+	struct sockaddr_in *broker4 =
+		((struct sockaddr_in *)&broker);
+
+	inet_pton(AF_INET, CONFIG_AZURE_IOT_HUB_STATIC_IPV4_ADDR,
+		  &broker->sin_addr);
+	broker4->sin_family = AF_INET;
+	broker4->sin_port = htons(CONFIG_AZURE_IOT_HUB_PORT);
+
+	LOG_DBG("IPv4 Address %s",
+		log_strdup(CONFIG_AZURE_IOT_HUB_STATIC_IPV4_ADDR));
+
+	return 0;
+}
+#else
+static int broker_init(bool dps)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo *addr;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM
+	};
+
+	if (dps) {
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+		err = getaddrinfo(CONFIG_AZURE_IOT_HUB_DPS_HOSTNAME,
+				  NULL, &hints, &result);
+		if (err) {
+			LOG_ERR("getaddrinfo, error %d", err);
+			return -ECHILD;
```

Comment:
I'm not sure what the "no children" error codes mean. Doesn't seem to fit here?

Reply:
Looks like it's commonly used in NCS, at least in all cloud backends, but can't see any consistency across the rest of the codebase. Anyway, I'll switch this to return the error code from `getaddrinfo`.